libs/vkd3d-shader/spirv.c

```diff
@@ -8514,6 +8521,39 @@ static uint32_t vkd3d_dxbc_compiler_adjust_ssbo_offset(struct vkd3d_dxbc_compile
     return coordinate_id;
 }
 
+static uint32_t vkd3d_dxbc_compiler_adjust_typed_buffer_offset(struct vkd3d_dxbc_compiler *compiler,
+        const struct vkd3d_shader_register *reg, uint32_t coordinate_id)
+{
+    struct vkd3d_spirv_builder *builder = &compiler->spirv_builder;
+    uint32_t bounds_id, offset_id, length_id;
+    uint32_t bool_type_id, uint_type_id;
+    uint32_t cond_id;
+    const struct vkd3d_symbol *symbol;
+
+    if (!(compiler->shader_interface.flags & VKD3D_SHADER_INTERFACE_TYPED_OFFSET_BUFFER))
+        return coordinate_id;
+
+    symbol = vkd3d_dxbc_compiler_find_resource(compiler, reg);
+    bounds_id = vkd3d_dxbc_compiler_get_buffer_bounds(compiler, reg, symbol->info.resource.resource_binding);
+
+    bool_type_id = vkd3d_spirv_get_type_id(builder, VKD3D_TYPE_BOOL, 1);
+    uint_type_id = vkd3d_spirv_get_type_id(builder, VKD3D_TYPE_UINT, 1);
+
+    offset_id = vkd3d_spirv_build_op_composite_extract1(builder, uint_type_id, bounds_id, 0);
+    length_id = vkd3d_spirv_build_op_composite_extract1(builder, uint_type_id, bounds_id, 1);
+
+    /* cond = offset < length */
+    cond_id = vkd3d_spirv_build_op_uless_than(builder, bool_type_id, coordinate_id, length_id);
+
+    /* In case of out-of-bounds access, set offset to a number that we
+     * expect to be out-of-bounds of the actual Vulkan resource as well. */
+    coordinate_id = vkd3d_spirv_build_op_select(builder, uint_type_id, cond_id,
+            vkd3d_spirv_build_op_iadd(builder, uint_type_id, coordinate_id, offset_id),
+            vkd3d_dxbc_compiler_get_constant_uint(compiler, 0xffffffffu));
```

Comment:
Do we know if this could potentially overflow and cause issues depending on view format?

Reply:
It passes the OOB tests on both AMD and NV. There cannot be address computation in the shader with typed views.